src/miral/runner.cpp

```diff
@@ -184,6 +205,31 @@ void miral::MirRunner::add_stop_callback(std::function<void()> const& stop_callb
     self->stop_callback = updated;
 }
 
+void miral::MirRunner::register_signal_handler(
+    std::initializer_list<int> signals,
+    std::function<void(int)> const& handler)
+{
+    std::lock_guard lock{self->mutex};
+    
+    if (auto const server = self->weak_server.lock())
+    {
+        auto const main_loop = server->the_main_loop();
+        main_loop->register_signal_handler(signals, handler);
+    }
+    else
+    {
+        auto const signal_info = Self::SignalInfo{signals, handler};
+        self->signal_backlog.push_back(signal_info);
+    }
+}
+
+auto miral::MirRunner::register_fd_handler(mir::Fd fd, std::function<void(int)> const& handler)
+-> std::unique_ptr<miral::FdHandle>
+{
+    std::lock_guard lock{self->mutex};
```

Comment:
Making calls to FdManager while holding the MirRunner mutex lock is a bad idea. This lock cannot fully protect FdManager (it won't be held when an FdHandle is dropped), and trying to lock multiple mutexes at the same time often leads to deadlocks. Since `fd_manager` ~~is~~ should be const, you don't need to hold a lock when calling it as long as it's threadsafe internally.